lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,361 @@
+-- MVar implementation and out of place.
+
+data Cert
+    = RegisterKey RewardAccount
+    | Delegate RewardAccount
+      -- ^ Which pool we're delegating to is here (and for now) irrelevant.
+      -- The main thing is that there exists a witness on-chain for this stake
+      -- key (registration certs don't require witnesses)
+    | DeRegisterKey RewardAccount
+    deriving (Eq, Show, Generic)
+
+--
+--
+-- a portfolio of 3 keys:
+-- [0, 1, 2]
+-- should have a nextKeyIx of 3
+--
+-- portfolio of 0 keys, should have a nextKeyIx of 0
+--
+-- FIXME: toEnum safety
+setPortfolioOf
+    :: (SoftDerivation k, ToRewardAccount k)
+    => DelegationState k
+    -> (k 'AddressK XPub -> Address)
+    -> (RewardAccount -> Bool) -- TODO: Need a Set or Map for the real implementation with LSQ
+    -> Int
+    -> Maybe Tx
+setPortfolioOf s mkAddress isReg n =
+    let s' = s { nextKeyIx = toEnum n }
+        mkTxIn (PointerUTxO i c) = (i, c)
+
+        minUTxOVal = Coin 1 -- FIXME
+        -- TODO: Need to rely on wallet to return as change, if the minUTxOVal
+        -- changes. Not sure if this is the case.
+        txWithCerts [] = Nothing
+        txWithCerts cs = Just $ Tx
+            { certs = cs
+            , inputs = maybeToList (mkTxIn <$> pointer s)
+            , outputs =
+                [ TxOut
+                    (mkAddress $ keyAtIx s $ nextKeyIx s')
+                    (TB.fromCoin minUTxOVal)
+                ]
+            }
+    in
+        txWithCerts $ reRegKey0IfNeeded
+            ++ deRegKey0IfNeeded
+            ++ case compare (toEnum n) (nextKeyIx s) of
+                GT -> deleg [nextKeyIx s .. toEnum (n - 1)]
+                EQ -> []
+                LT -> dereg $ reverse [toEnum n .. (pred $ nextKeyIx s)]
+  where
+    deleg = (>>= \ix ->
+        if isReg (acct ix)
+        then [Delegate (acct ix)]
+        else [RegisterKey (acct ix),  Delegate (acct ix)]
+        )
+
+    reRegKey0IfNeeded =
+        if not (isKey0Reg s) && n > 0 && fromEnum (nextKeyIx s) > 1
+        then [RegisterKey (acct minBound), Delegate (acct minBound)]
+        else []
+
+    deRegKey0IfNeeded =
+        [ DeRegisterKey (acct minBound)
+        | (isKey0Reg s)
+        , n == 0
+        , fromEnum (nextKeyIx s) == 0
+        ]
+
+    dereg ixs =
+        [ DeRegisterKey (acct ix)
+        | ix <- ixs
+        , isReg . acct $ ix
+        -- We need to /at least/ check @isReg (key 0)@, because the user could
+        -- have deregistered it using old wallet software.
+        ]
+
+    acct = toRewardAccount . keyAtIx s
+
+applyTx
+    :: forall k. ( SoftDerivation k
+        , ToRewardAccount k
+        , MkKeyFingerprint k Address
+        , MkKeyFingerprint k (k 'AddressK XPub))
+    => Tx
+    -> Hash "Tx"
+    -> DelegationState k
+    -> DelegationState k
+applyTx (Tx cs _ins outs) h s0 =
+    let
+        s = foldl (flip applyCert) s0 cs
+        isOurOut i (TxOut addr _b) =
```

Comment:
👍 I changed it to assume the pointer _only_ changes based on the delegation certificates.

> The CIP explicitly requires that the first output of a such a transaction is the pointer output

Wouldn't that be an unnecessary loss of flexibility? 🤔 (Not that I know what we'd use it for, maybe changing portfolio of multiple accounts at the same time 🤷‍♂️ )

Reply:
> Wouldn't that be an unnecessary loss of flexibility? thinking (Not that I know what we'd use it for, maybe changing portfolio of multiple accounts at the same time man_shrugging )

More of a convention (perhaps a useless one?). In the end, since the wallet tracks the UTxO it's quite easy to locate it in the input set.